// @pulumi/aws/bedrock/index.js
"use strict";
// *** WARNING: this file was generated by the Pulumi Terraform Bridge (tfgen) Tool. ***
// *** Do not edit by hand unless you're certain you know what you are doing! ***
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProvisionedModelThroughput = exports.getCustomModelsOutput = exports.getCustomModels = exports.getCustomModelOutput = exports.getCustomModel = exports.CustomModel = exports.AgentKnowledgeBase = exports.AgentDataSource = exports.AgentAgentKnowledgeBaseAssociation = exports.AgentAgentAlias = exports.AgentAgentActionGroup = exports.AgentAgent = void 0;
const pulumi = require("@pulumi/pulumi");
const utilities = require("../utilities");
exports.AgentAgent = null;
utilities.lazyLoad(exports, ["AgentAgent"], () => require("./agentAgent"));
exports.AgentAgentActionGroup = null;
utilities.lazyLoad(exports, ["AgentAgentActionGroup"], () => require("./agentAgentActionGroup"));
exports.AgentAgentAlias = null;
utilities.lazyLoad(exports, ["AgentAgentAlias"], () => require("./agentAgentAlias"));
exports.AgentAgentKnowledgeBaseAssociation = null;
utilities.lazyLoad(exports, ["AgentAgentKnowledgeBaseAssociation"], () => require("./agentAgentKnowledgeBaseAssociation"));
exports.AgentDataSource = null;
utilities.lazyLoad(exports, ["AgentDataSource"], () => require("./agentDataSource"));
exports.AgentKnowledgeBase = null;
utilities.lazyLoad(exports, ["AgentKnowledgeBase"], () => require("./agentKnowledgeBase"));
exports.CustomModel = null;
utilities.lazyLoad(exports, ["CustomModel"], () => require("./customModel"));
exports.getCustomModel = null;
exports.getCustomModelOutput = null;
utilities.lazyLoad(exports, ["getCustomModel", "getCustomModelOutput"], () => require("./getCustomModel"));
exports.getCustomModels = null;
exports.getCustomModelsOutput = null;
utilities.lazyLoad(exports, ["getCustomModels", "getCustomModelsOutput"], () => require("./getCustomModels"));
exports.ProvisionedModelThroughput = null;
utilities.lazyLoad(exports, ["ProvisionedModelThroughput"], () => require("./provisionedModelThroughput"));
const _module = {
    version: utilities.getVersion(),
    construct: (name, type, urn) => {
        switch (type) {
            case "aws:bedrock/agentAgent:AgentAgent":
                return new exports.AgentAgent(name, undefined, { urn });
            case "aws:bedrock/agentAgentActionGroup:AgentAgentActionGroup":
                return new exports.AgentAgentActionGroup(name, undefined, { urn });
            case "aws:bedrock/agentAgentAlias:AgentAgentAlias":
                return new exports.AgentAgentAlias(name, undefined, { urn });
            case "aws:bedrock/agentAgentKnowledgeBaseAssociation:AgentAgentKnowledgeBaseAssociation":
                return new exports.AgentAgentKnowledgeBaseAssociation(name, undefined, { urn });
            case "aws:bedrock/agentDataSource:AgentDataSource":
                return new exports.AgentDataSource(name, undefined, { urn });
            case "aws:bedrock/agentKnowledgeBase:AgentKnowledgeBase":
                return new exports.AgentKnowledgeBase(name, undefined, { urn });
            case "aws:bedrock/customModel:CustomModel":
                return new exports.CustomModel(name, undefined, { urn });
            case "aws:bedrock/provisionedModelThroughput:ProvisionedModelThroughput":
                return new exports.ProvisionedModelThroughput(name, undefined, { urn });
            default:
                throw new Error(`unknown resource type ${type}`);
        }
    },
};
pulumi.runtime.registerResourceModule("aws", "bedrock/agentAgent", _module);
pulumi.runtime.registerResourceModule("aws", "bedrock/agentAgentActionGroup", _module);
pulumi.runtime.registerResourceModule("aws", "bedrock/agentAgentAlias", _module);
pulumi.runtime.registerResourceModule("aws", "bedrock/agentAgentKnowledgeBaseAssociation", _module);
pulumi.runtime.registerResourceModule("aws", "bedrock/agentDataSource", _module);
pulumi.runtime.registerResourceModule("aws", "bedrock/agentKnowledgeBase", _module);
pulumi.runtime.registerResourceModule("aws", "bedrock/customModel", _module);
pulumi.runtime.registerResourceModule("aws", "bedrock/provisionedModelThroughput", _module);
